src/rest/class-wp-rest-abilities-run-controller.php

```diff
@@ -0,0 +1,311 @@
+					'methods'             => WP_REST_Server::ALLMETHODS,
+					'callback'            => array( $this, 'run_ability_with_method_check' ),
+					'permission_callback' => array( $this, 'run_ability_permissions_check' ),
+					'args'                => $this->get_run_args(),
+				),
+				'schema' => array( $this, 'get_run_schema' ),
+			)
+		);
+	}
+
+	/**
+	 * Executes an ability with HTTP method validation.
+	 *
+	 * @since 0.1.0
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 */
+	public function run_ability_with_method_check( $request ) {
+		$ability = wp_get_ability( $request['id'] );
+
+		if ( ! $ability ) {
+			return new WP_Error(
+				'rest_ability_not_found',
+				__( 'Ability not found.' ),
+				array( 'status' => 404 )
+			);
+		}
+
+		// Check if the HTTP method matches the ability type
+		$meta = $ability->get_meta();
+		$type = isset( $meta['type'] ) ? $meta['type'] : 'tool';
+		$method = $request->get_method();
+
+		if ( 'resource' === $type && 'GET' !== $method ) {
+			return new WP_Error(
+				'rest_invalid_method',
+				__( 'Resource abilities require GET method.' ),
+				array( 'status' => 405 )
+			);
+		} elseif ( 'resource' !== $type && 'POST' !== $method ) {
+			return new WP_Error(
+				'rest_invalid_method',
+				__( 'Tool abilities require POST method.' ),
+				array( 'status' => 405 )
+			);
+		}
+
+		return $this->run_ability( $request );
+	}
+
+	/**
+	 * Executes an ability.
+	 *
+	 * @since 0.1.0
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 */
+	public function run_ability( $request ) {
+		$ability = wp_get_ability( $request['id'] );
+
+		if ( ! $ability ) {
+			return new WP_Error(
+				'rest_ability_not_found',
+				__( 'Ability not found.' ),
+				array( 'status' => 404 )
+			);
+		}
+
+		if ( 'GET' === $request->get_method() ) {
+			$input = $request->get_query_params();
+			unset( $input['_locale'], $input['context'], $input['page'], $input['per_page'] );
+		} else {
+			$input = $request->get_json_params() ?: array();
+		}
+
+		$input_validation = $this->validate_input( $ability, $input );
+		if ( is_wp_error( $input_validation ) ) {
+			return $input_validation;
+		}
+
+		$result = $ability->execute( $input );
+
+		if ( is_wp_error( $result ) ) {
+			return new WP_Error(
+				'rest_ability_execution_failed',
+				$result->get_error_message(),
+				array( 'status' => 500 )
+			);
+		}
+
+		if ( is_null( $result ) ) {
+			return new WP_Error(
+				'rest_ability_execution_failed',
+				__( 'Ability execution failed. Please check permissions and input parameters.' ),
+				array( 'status' => 500 )
+			);
+		}
+
+		$output_validation = $this->validate_output( $ability, $result );
+		if ( is_wp_error( $output_validation ) ) {
+			return $output_validation;
+		}
```

Comment:
This part is also not needed, as it is implemented inside `$ability->execute( $input )`.